### ISSUE TITLE
server: packet integrity error when ctx.accept() publickey request

### ISSUE THREAD

**Issue by mofux** [mscdex/ssh2#259]
Using the server example from the README.md and testing publickey auth,
I get the following error when I `ctx.accept()` in the authentication handler (ctx.signature is not present):

```
OpenSSH_6.2p2, OSSLShim 0.9.8r 8 Dec 2011
debug1: Reading configuration data /etc/ssh_config
debug1: /etc/ssh_config line 20: Applying options for *
debug2: ssh_connect: needpriv 0
debug1: Connecting to localhost [::1] port 8080.
debug1: connect to address ::1 port 8080: Connection refused
debug1: Connecting to localhost [127.0.0.1] port 8080.
debug1: Connection established.
debug3: Incorrect RSA1 identifier
debug3: Could not load "/Users/user/.ssh/id_rsa" as a RSA1 public key
debug1: identity file /Users/user/.ssh/id_rsa type 1
debug1: identity file /Users/user/.ssh/id_rsa-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_6.2
debug1: Remote protocol version 2.0, remote software version ssh2js0.0.9srv
debug1: no match: ssh2js0.0.9srv
debug2: fd 5 setting O_NONBLOCK
debug3: put_host_port: [localhost]:8080
debug3: load_hostkeys: loading entries for host "[localhost]:8080" from file "/Users/user/.ssh/known_hosts"
debug3: load_hostkeys: found key type RSA in file /Users/user/.ssh/known_hosts:7
debug3: load_hostkeys: loaded 1 keys
debug3: order_hostkeyalgs: prefer hostkeyalgs: ssh-rsa-cert-v01@openssh.com,ssh-rsa-cert-v00@openssh.com,ssh-rsa
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug2: kex_parse_kexinit: diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
debug2: kex_parse_kexinit: ssh-rsa-cert-v01@openssh.com,ssh-rsa-cert-v00@openssh.com,ssh-rsa,ssh-dss-cert-v01@openssh.com,ssh-dss-cert-v00@openssh.com,ssh-dss
debug2: kex_parse_kexinit: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
debug2: kex_parse_kexinit: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
debug2: kex_parse_kexinit: hmac-md5-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-ripemd160-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-96-etm@openssh.com,hmac-md5,hmac-sha1,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
debug2: kex_parse_kexinit: hmac-md5-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-ripemd160-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-96-etm@openssh.com,hmac-md5,hmac-sha1,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
debug2: kex_parse_kexinit: none,zlib@openssh.com,zlib
debug2: kex_parse_kexinit: none,zlib@openssh.com,zlib
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: first_kex_follows 0 
debug2: kex_parse_kexinit: reserved 0 
debug2: kex_parse_kexinit: diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
debug2: kex_parse_kexinit: ssh-rsa,ssh-dss
debug2: kex_parse_kexinit: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
debug2: kex_parse_kexinit: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
debug2: kex_parse_kexinit: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
debug2: kex_parse_kexinit: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
debug2: kex_parse_kexinit: none
debug2: kex_parse_kexinit: none
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: first_kex_follows 0 
debug2: kex_parse_kexinit: reserved 0 
debug2: mac_setup: found hmac-md5
debug1: kex: server->client aes128-ctr hmac-md5 none
debug2: mac_setup: found hmac-md5
debug1: kex: client->server aes128-ctr hmac-md5 none
debug2: dh_gen_key: priv key bits set: 138/256
debug2: bits set: 1018/2048
debug1: sending SSH2_MSG_KEXDH_INIT
debug1: expecting SSH2_MSG_KEXDH_REPLY
debug1: Server host key: RSA f6:fa:d3:0d:de:3e:3b:2e:45:55:87:ef:18:4d:77:29
debug3: put_host_port: [127.0.0.1]:8080
debug3: put_host_port: [localhost]:8080
debug3: load_hostkeys: loading entries for host "[localhost]:8080" from file "/Users/user/.ssh/known_hosts"
debug3: load_hostkeys: found key type RSA in file /Users/user/.ssh/known_hosts:7
debug3: load_hostkeys: loaded 1 keys
debug1: Host '[localhost]:8080' is known and matches the RSA host key.
debug1: Found key in /Users/user/.ssh/known_hosts:7
debug2: bits set: 1054/2048
debug1: ssh_rsa_verify: signature correct
debug2: kex_derive_keys
debug2: set_newkeys: mode 1
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug2: set_newkeys: mode 0
debug1: SSH2_MSG_NEWKEYS received
debug1: Roaming not allowed by server
debug1: SSH2_MSG_SERVICE_REQUEST sent
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug2: key: /Users/user/.ssh/id_rsa (0x7f97fae00150), explicit
debug1: Authentications that can continue: 
debug3: start over, passed a different list publickey,keyboard-interactive,password
debug3: preferred publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Offering RSA public key: /Users/user/.ssh/id_rsa
debug3: send_pubkey_test
debug2: we sent a publickey packet, wait for reply
Packet integrity error (272 bytes remaining) at /SourceCache/OpenSSH/OpenSSH-189/openssh/sshconnect2.c:642
Disconnecting: Packet integrity error.
```

and for the server debug log:

```
[60266.666241142] DEBUG: Local ident: 'SSH-2.0-ssh2js0.0.9srv'
[60266.666241142] DEBUG: Parser: IN_INIT (waiting for identification begin)
[60266.666241142] DEBUG: Parser: IN_INIT (waiting for rest of identification)
[60266.666241142] DEBUG: Parser: IN_GREETING
[60266.666241142] DEBUG: Parser: IN_HEADER
[60266.666241142] DEBUG: Remote ident: 'SSH-2.0-OpenSSH_6.2'
[60266.666241142] DEBUG: Outgoing: Writing KEXINIT
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: pktLen:1588,padLen:6,remainLen:1584
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
[60266.666241142] DEBUG: Comparing KEXINITs ...
[60266.666241142] DEBUG: (local) KEX algorithms: diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
[60266.666241142] DEBUG: (remote) KEX algorithms: diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
[60266.666241142] DEBUG: KEX algorithm: diffie-hellman-group14-sha1
[60266.666241142] DEBUG: (local) Host key formats: ssh-rsa,ssh-dss
[60266.666241142] DEBUG: (remote) Host key formats: ssh-rsa-cert-v01@openssh.com,ssh-rsa-cert-v00@openssh.com,ssh-rsa,ssh-dss-cert-v01@openssh.com,ssh-dss-cert-v00@openssh.com,ssh-dss
[60266.666241142] DEBUG: Host key format: ssh-rsa
[60266.666241142] DEBUG: (local) Client->Server ciphers: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
[60266.666241142] DEBUG: (remote) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
[60266.666241142] DEBUG: Client->Server Cipher: aes128-ctr
[60266.666241142] DEBUG: (local) Server->Client ciphers: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
[60266.666241142] DEBUG: (remote) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
[60266.666241142] DEBUG: Server->Client Cipher: aes128-ctr
[60266.666241142] DEBUG: (local) Client->Server HMAC algorithms: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
[60266.666241142] DEBUG: (remote) Client->Server HMAC algorithms: hmac-md5-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-ripemd160-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-96-etm@openssh.com,hmac-md5,hmac-sha1,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
[60266.666241142] DEBUG: Client->Server HMAC algorithm: hmac-md5
[60266.666241142] DEBUG: (local) Server->Client HMAC algorithms: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
[60266.666241142] DEBUG: (remote) Server->Client HMAC algorithms: hmac-md5-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-ripemd160-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-96-etm@openssh.com,hmac-md5,hmac-sha1,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
[60266.666241142] DEBUG: Server->Client HMAC algorithm: hmac-md5
[60266.666241142] DEBUG: (local) Client->Server compression algorithms: none
[60266.666241142] DEBUG: (remote) Client->Server compression algorithms: none,zlib@openssh.com,zlib
[60266.666241142] DEBUG: Client->Server compression algorithm: none
[60266.666241142] DEBUG: (local) Server->Client compression algorithms: none
[60266.666241142] DEBUG: (remote) Server->Client compression algorithms: none,zlib@openssh.com,zlib
[60266.666241142] DEBUG: Server->Client compression algorithm: none
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: pktLen:268,padLen:6,remainLen:264
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_INIT
[60266.666241142] DEBUG: Outgoing: Writing KEXDH_REPLY
[60266.666241142] DEBUG: Outgoing: Writing NEWKEYS
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: pktLen:12,padLen:10,remainLen:8
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: NEWKEYS
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: HMAC size:16
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY
[60266.666241142] DEBUG: Parser: Verifying MAC
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: SERVICE_REQUEST
[60266.666241142] DEBUG: Outgoing: Writing SERVICE_ACCEPT (ssh-userauth)
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: pktLen:60,padLen:6,remainLen:48
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: HMAC size:16
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY
[60266.666241142] DEBUG: Parser: Verifying MAC
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_REQUEST (none)
[60266.666241142] DEBUG: Outgoing: Writing USERAUTH_FAILURE
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: pktLen:364,padLen:10,remainLen:352
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: HMAC size:16
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY
[60266.666241142] DEBUG: Parser: Verifying MAC
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_REQUEST (publickey)
[60266.666241142] DEBUG: Outgoing: Writing USERAUTH_PK_OK
[60266.666241142] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[60266.666241142] DEBUG: Parser: IN_PACKET
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: pktLen:44,padLen:7,remainLen:32
[60266.666241142] DEBUG: Parser: IN_PACKETDATA
[60266.666241142] DEBUG: Parser: Decrypting
[60266.666241142] DEBUG: Parser: HMAC size:16
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY
[60266.666241142] DEBUG: Parser: Verifying MAC
[60266.666241142] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[60266.666241142] DEBUG: Parser: IN_PACKETDATAAFTER, packet: DISCONNECT (PROTOCOL_ERROR)
```
I am using node 0.12.2 but had the same error with 0.12.0

Is the example still working for you?

**Comment by mofux** [comment]
Here is some code that can be used to reproduce the problem:

```javascript
var ssh2 = require('ssh2');
var fs = require('fs');
var privateKey = '/home/vagrant/.ssh/id_rsa';

var server = new ssh2.Server({ privateKey: fs.readFileSync(privateKey), debug: console.log }, function(client) {

	client.on('authentication', function(ctx) {

		if (ctx.method === 'publickey') {
			ctx.accept();
		} else {
			ctx.reject();
		}

	});

	client.on('ready', function() {

		console.log('client authenticated');
		client.on('session', function(accept, reject) {
			accept();
		});

	});

	client.on('error', console.log);

});

server.on('error', console.log);

server.listen(8080);
```

**Comment by mscdex** [comment]
What module version are you using?

**Comment by mofux** [comment]
The current master branch, my package.json links against `mscdex/ssh2` which currently is 0.4.7 if I check the package.json of the module.

**Comment by mofux** [comment]
Btw, here is the debug output of a successful ssh connection using publickey auth against an OpenSSH server:

```
debug2: kex_parse_kexinit: none,zlib@openssh.com
debug2: kex_parse_kexinit: none,zlib@openssh.com
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: 
debug2: kex_parse_kexinit: first_kex_follows 0 
debug2: kex_parse_kexinit: reserved 0 
debug2: mac_setup: setup hmac-md5-etm@openssh.com
debug1: kex: server->client aes128-ctr hmac-md5-etm@openssh.com none
debug2: mac_setup: setup hmac-md5-etm@openssh.com
debug1: kex: client->server aes128-ctr hmac-md5-etm@openssh.com none
debug1: sending SSH2_MSG_KEX_ECDH_INIT
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ECDSA 03:ce:69:e9:9a:55:89:ed:f0:53:b5:f2:e9:74:53:a9
debug3: load_hostkeys: loading entries for host "localhost" from file "/home/vagrant/.ssh/known_hosts"
debug3: load_hostkeys: found key type ECDSA in file /home/vagrant/.ssh/known_hosts:4
debug3: load_hostkeys: loaded 1 keys
debug1: Host 'localhost' is known and matches the ECDSA host key.
debug1: Found key in /home/vagrant/.ssh/known_hosts:4
debug1: ssh_ecdsa_verify: signature correct
debug2: kex_derive_keys
debug2: set_newkeys: mode 1
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug2: set_newkeys: mode 0
debug1: SSH2_MSG_NEWKEYS received
debug1: Roaming not allowed by server
debug1: SSH2_MSG_SERVICE_REQUEST sent
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug2: key: /home/vagrant/.ssh/id_rsa (0x7f0711d3ccf0),
debug2: key: /home/vagrant/.ssh/id_dsa (0x7f0711d412a0),
debug2: key: /home/vagrant/.ssh/id_ecdsa ((nil)),
debug2: key: /home/vagrant/.ssh/id_ed25519 ((nil)),
debug1: Authentications that can continue: publickey,password
debug3: start over, passed a different list publickey,password
debug3: preferred gssapi-keyex,gssapi-with-mic,publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Offering RSA public key: /home/vagrant/.ssh/id_rsa
debug3: send_pubkey_test
debug2: we sent a publickey packet, wait for reply
debug1: Server accepts key: pkalg ssh-rsa blen 279
debug2: input_userauth_pk_ok: fp f6:fa:d3:0d:de:3e:3b:2e:45:55:87:ef:18:4d:77:29
debug3: sign_and_send_pubkey: RSA f6:fa:d3:0d:de:3e:3b:2e:45:55:87:ef:18:4d:77:29
debug1: key_parse_private2: missing begin marker
debug1: read PEM private key done: type RSA
debug1: Authentication succeeded (publickey).
... more session logs ...
```

**Comment by mofux** [comment]
I think I have found the problem. In `ssh2-streams/lib/ssh.js` in function `SSH2Stream.prototype.authPKOK ` the length of the key is not added to the buffer, causing the error. Below is the fixed function adding the extra 4 bytes for the length of the keyLen:

```javascript
SSH2Stream.prototype.authPKOK = function(keyAlgo, key) {
  if (!this.server)
    throw new Error('Server-only method called in client mode');

  var keyAlgoLen = keyAlgo.length,
      keyLen = key.length,
      buf = new Buffer(1 + 4 + keyAlgoLen + 4 + keyLen);

  buf[0] = MESSAGE.USERAUTH_PK_OK;

  buf.writeUInt32BE(keyAlgoLen, 1, true);
  buf.write(keyAlgo, 5, keyAlgoLen, 'utf8');

  buf.writeUInt32BE(keyLen, 5 + keyAlgoLen, true);
  key.copy(buf, 5 + keyAlgoLen + 4);

  this.debug('DEBUG: Outgoing: Writing USERAUTH_PK_OK');
  return send(this, buf);
};
```

**Comment by mscdex** [comment]
@mofux Thanks, you're right. This should now be fixed in cedb3b8b40.
